art/attacks/evasion/sign_opt.py

```diff
@@ -0,0 +1,561 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+This module implements the Sign-OPT attack `SignOPTAttack`. This is a query-efficient
+hard-label adversarial attack.
+| Paper link: https://arxiv.org/pdf/1909.10773.pdf
+"""
+
+import logging
+from typing import Optional, TYPE_CHECKING, Tuple
+import time
+import numpy as np
+from numpy import integer, linalg as LA
+from tqdm.auto import tqdm
+from art.attacks.attack import EvasionAttack
+from art.config import ART_NUMPY_DTYPE
+from art.estimators.estimator import BaseEstimator
+from art.estimators.classification.classifier import ClassifierMixin
+from art.utils import compute_success, check_and_transform_label_format, get_labels_np_array
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class SignOPTAttack(EvasionAttack):
+    """
+    Implements the Sign-OPT attack `SignOPTAttack`. This is a query-efficient
+    hard-label adversarial attack.
+    | Paper link: https://arxiv.org/pdf/1909.10773.pdf
+    """
+
+    attack_params = EvasionAttack.attack_params + [
+        "targeted",
+        "epsilon",
+        "num_trial",
+        "max_iter",
+        "query_limit",
+        "K",
+        "alpha",
+        "beta",
+        "batch_size",
+        "verbose",
+    ]
+
+    _estimator_requirements = (BaseEstimator, ClassifierMixin)
+
+    def __init__(
+        self,
+        estimator: "CLASSIFIER_TYPE",
+        targeted: bool = True,
+        epsilon: float = 0.001,
+        num_trial: int = 100,
+        max_iter: int = 1000,  # recommend 5000 for targeted attack
+        query_limit: int = 20000,  # recommend 40000 for targeted attack
+        k: int = 200,
+        alpha: float = 0.2,
+        beta: float = 0.001,
+        eval_perform: bool = False,
+        batch_size: int = 64,
+        verbose: bool = False,
+    ) -> None:
+        """
+        Create a Sign_OPT attack instance.
+        :param estimator: A trained classifier.
```

Comment:
An empty line is required after the description and before the param section for the documentation to build correctly.
```suggestion
        Create a Sign_OPT attack instance.

        :param estimator: A trained classifier.
```